core/corehttp/proxy.go

```diff
@@ -0,0 +1,132 @@
+package corehttp
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+
+	core "github.com/ipfs/go-ipfs/core"
+
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+	peer "gx/ipfs/QmbNepETomvmXfz1X5pHNFD2QuPqnqi47dTd94QJWSorQ3/go-libp2p-peer"
+	inet "gx/ipfs/QmfDPh144WGBqRxZb1TGDHerbMnZATrHZggAPw7putNnBq/go-libp2p-net"
+)
+
+// ProxyOption is an endpoint for proxying a HTTP request to another ipfs peer
+func ProxyOption() ServeOption {
+	return func(ipfsNode *core.IpfsNode, _ net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
+		mux.HandleFunc("/proxy/http/", func(w http.ResponseWriter, request *http.Request) {
+			// parse request
+			parsedRequest, err := parseRequest(request)
+			if err != nil {
+				handleError(w, "Failed to parse request", err, 400)
+				return
+			}
+
+			// open connect to peer
+			stream, err := ipfsNode.P2P.PeerHost.NewStream(request.Context(), parsedRequest.target, protocol.ID("/x/"+parsedRequest.name))
+			if err != nil {
+				msg := fmt.Sprintf("Failed to open stream '%v' to target peer '%v'", parsedRequest.name, parsedRequest.target)
+				handleError(w, msg, err, 500)
+				return
+			}
+			//send proxy request and response to client
+			newReverseHTTPProxy(parsedRequest, stream).ServeHTTP(w, request)
+		})
+		return mux, nil
+	}
+}
+
+type proxyRequest struct {
+	target   peer.ID
+	name     string
+	httpPath string // path to send to the proxy-host
+}
+
+// from the url path parse the peer-ID, name and http path
+// /proxy/http/$peer_id/$name/$http_path
+func parseRequest(request *http.Request) (*proxyRequest, error) {
+	path := request.URL.Path
+
+	split := strings.SplitN(path, "/", 6)
```

Comment:
This is going to be problematic with the way protocol tags usually look like because normally they have one or several slashes `/x/something`. I am not sure if this is a spec requirement regarding protocol IDs or something.

I was thinking this would be able to reach the cluster libp2p-http API but it uses `/libp2p-http` for protocol tag, and it's not possible to express this as part of the url path (easily).

I would suggest to settle on a default protocol tag for proxying http and allowing modifications via a query-parameter (or a custom header).

Reply:
> I would suggest to settle on a default protocol tag for proxying http and allowing modifications via a query-parameter

Unfortunately, we can't use a query parameter (as we can't distinguish between ones that need to be forwarded). We *could* use a header but that's not very user/browser-friendly.

We currently use `/x/something` because `ipfs p2p` listens on `/x/...` protocols (by default). The idea behind this feature was to allow a client to connect to `http://localhost:5001/proxy/http/QmId/something` and have this entire request forwarded to some remote `ipfs p2p` listener. We could just always use the same protocol but I really want to be able to use IPFS to proxy to arbitrary webservers.

On the other hand, being able to connect to ipfs-cluster would be awesome. Additionally, I can see us eventually adding a custom protocol to browsers.

We *could* just encode them.

---

An alternative is to make the *server* smarter. That is, we can:

1. Listen on one protocol (possibly allowing the protocol to be set via a header).
2. Add a new `ipfs http` command.

The user would then run `ipfs http route add / http://localhost:1224/a/b` to, e.g., route all traffic to that webserver. The upside is:

1. This server can add authenticated `x-libp2p-peer-id: QmClient` headers.
2. Flexible routes.
3. We can even serve plain files/directories (act like a gateway).

However, this is starting to look like a *big* feature. Really, we I think we can probably get away with both (although we may still want some encoding scheme to allow *this* feature to talk to cluster, thoughts)?

That is, we can eventually add a single-protocol forwarder at `http://gateway/http/QmId/`.

Thoughts?

Reply:
Would it be that restrictive to not allow /'s in protocol names? Logically it isn't and I suspect this issue will otherwise crop up in other places with URLs using protocol names..

Reply:
IMO, not really. However, it *does* restrict this command to `/x/` protocols. I'm just polling for feedback and trying to make sure we don't paint ourselves into a corner.

Regerdless, I don't think this discussion should block this PR. This *is* an experimental API so, if we have a good enough reason, we can still change it.

Reply:
If you made that prefix "x-" then even that restriction would go away as well.

Reply:
I think just as you can have a custom header you could have a custom query parameter `?x-fw-protocol=/x/test` which you strip (I realize this is not pretty but it's so easy to use).

Altenatively, the best is probably url encoding the slashes. That's what encoding is for anyway, but it might give some headaches to users at some point.

I'm not sure the added complexity making server smarter compensates that right now it's quite simple and solves the proposing user's usecase. But in that case `route add` could take a custom protocol tag for the route too.

Reply:
> I think just as you can have a custom header you could have a custom query parameter `?x-fw-protocol=/x/test` which you strip (I realize this is not pretty but it's so easy to use).

This is just a bug waiting to happen. I guess we could do `/proxy/http/Qm...?protocol=xyz&path=/a/b/c`.

> If you made that prefix "x-" then even that restriction would go away as well.

To be consistent with pretty much everything else we do, I'd like to stick to `/x/`. For context, "protocols" were envisioned to be `/ipfs/QmFile/path/to/spec`. My suggestion with `x-...` was that we could encode `/x/...` as `x-...` in the url.

Reply:
The main issue I see with a custom header, or query parameter is you lose transitivity. i.e. I can't pipe a query over a p2p stream to one peer, which then pipes it to another peer. But maybe that's a weird use case..

I think my vote would be for url encoding the protocol name. 

Encoding the path in the query parameter is a no go because then you lose the fundamental feature of a a website "just working" behind a p2p proxy.

Reply:
> My suggestion with x-... was that we could encode /x/... as x-... in the url

So cluster uses `/libp2p-http` for protocol, so if you want it to work you cannot use that encoding. I say, let's use a default protocol and put the custom one in a header?. It's not obvious to a user that he should write  `-a-b-c` in one place and then listen to protocol `/a/b/c` in some other.

Reply:
What do we think about url encoding the protocol where it is currently, in the path? Headers won't work because a website served through this won't add the magic header to sub resource requests. Query parameters have the same issue.